blog/content/post/2022-01-03-pcie_screamer.md

```diff
@@ -0,0 +1,218 @@
+a correct design that can be programmed into the device. There are several
+projects that implement example designs for PCIeScreamer that can be later
+modified to meet our needs.
+
+For this usage, the
+[enjoy-digital/pcie_screamer](https://github.com/enjoy-digital/pcie_screamer)
+project was chosen. However, despite the simple instruction on the project
+[README](https://github.com/enjoy-digital/pcie_screamer) page, it is not very
+straightforward for first use.
+
+To prepare `enjoy-digital/pcie_screamer`, the following steps were tried:
+
+1. Install Vivado.
+1. Add Vivado to the `$PATH` environment variable.
+
+   ```
+   export PATH=$PATH:~/Xilinx/Vivado/2021.2/bin
+   ```
+
+   Your Vivado location may be different depending on options chosen during the
+   installation process and installed version.
+
+   You can check if you have done it correctly using the following command:
+
+   ```
+   vivado -version
+   ```
+
+1. Get and configure OpenCOD source code
+
+   ```
+   git clone https://github.com/openocd-org/openocd
+   cd openocd
+
+   git fetch https://review.openocd.org/openocd refs/changes/01/6801/1 && git checkout FETCH_HEAD
+
+   ./bootstrap
+   ./configure
+   ```
+
+1. Prepare a separate folder for files so the scripts won't mess inside your
+   important directory and change your current working directory to it.
+1. Clone
+   [enjoy-digital/pcie_screamer](https://github.com/enjoy-digital/pcie_screamer)
+   and checkout the correct branch
+
+   ```
+   git clone git@github.com:3mdeb/pcie_screamer.git
+   cd pcie_screamer
+   git checkout 01aee32f54f15163fac870cda3739ddd77b3c03c
+   ```
+
+2. Prepare [enjoy-digital/litex](https://github.com/enjoy-digital/litex)
+   environment
+
+   ```
+   wget https://raw.githubusercontent.com/enjoy-digital/litex/master/litex_setup.py
+   chmod +x litex_setup.py
+   ./litex_setup.py --init --install --user
+   ```
+
+   > This step will extract plenty of files in the directory, be sure to execute
+   > it from the `pcie_screamer` project directory.
+
+3. Now, the building of `Xilinx BIT data` is working
+
+   ```
+   ./pcie_screamer.py --build
+   ```
+
+4. However, the loading fails
+
+   ```
+   $ ./pcie_screamer.py --load
+
+   <output truncated>
+
+   failed loading file build/gateware/top.bit to pld device 0
+   Traceback (most recent call last):
+     File "./pcie_screamer.py", line 160, in <module>
+       main()
+     File "./pcie_screamer.py", line 150, in main
+       prog.load_bitstream("build/gateware/top.bit")
+     File "/home/ibagnucki/git/pcie_screamer/litex/litex/litex/build/openocd.py", line 27, in load_bitstream
+       self.call(["openocd", "-f", config, "-c", script])
+     File "/home/ibagnucki/git/pcie_screamer/litex/litex/litex/build/generic_programmer.py", line 100, in call
+       raise OSError(msg)
+   OSError: Error occured during OpenOCD's call, please check:
+   - OpenOCD installation.
+   - access permissions.
+   - hardware and cable.
```

Comment:
Having JTAG header soldered on PCIScreamer board one can use JTAG programmer for Xilinx FPGAs (best Xilinx Cable) for uploading binary configuration file into board. For this purpose "Vivado Hardware Manager" can be used. As far as I understand configuration file for FPGA was created in earlier step.